packages/integrations/markdoc/src/load-config.ts

```diff
@@ -0,0 +1,101 @@
+import type { AstroConfig } from 'astro';
+import type { Config as MarkdocConfig } from '@markdoc/markdoc';
+import { build as esbuild } from 'esbuild';
+import { fileURLToPath } from 'node:url';
+import * as fs from 'node:fs';
+
+const SUPPORTED_MARKDOC_CONFIG_FILES = [
+	'markdoc.config.js',
+	'markdoc.config.mjs',
+	'markdoc.config.ts',
+];
+
+export async function loadMarkdocConfig(astroConfig: Pick<AstroConfig, 'root'>) {
+	let markdocConfigUrl: URL | undefined;
+	for (const filename of SUPPORTED_MARKDOC_CONFIG_FILES) {
+		const filePath = new URL(filename, astroConfig.root);
+		if (!fs.existsSync(filePath)) continue;
+
+		markdocConfigUrl = filePath;
+		break;
+	}
+	if (!markdocConfigUrl) return;
+
+	const { code, dependencies } = await bundleConfigFile({
+		markdocConfigUrl,
+		astroConfig,
+	});
+	const config: MarkdocConfig = await loadConfigFromBundledFile(astroConfig.root, code);
+
+	return {
+		config,
+		fileUrl: markdocConfigUrl,
+	};
+}
+
+/**
+ * Forked from Vite's `bundleConfigFile` function
+ * with added handling for `.astro` imports,
+ * and removed unused Deno patches.
+ * @see https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts#L961
+ */
+async function bundleConfigFile({
```

Comment:
👀  I think it would be great if we can prevent using esbuild for the config. IIUC the markdoc config is evaluated in node so it can be validated before rendering: https://github.com/withastro/astro/blob/b75570d34ca0b528874b255b791df58175524b2e/packages/integrations/markdoc/src/index.ts#L56-L61

Could we add a validation call in `Renderer.astro` instead? https://github.com/withastro/astro/blob/b75570d34ca0b528874b255b791df58175524b2e/packages/integrations/markdoc/components/Renderer.astro#L13-L14

That way the config is only loaded as a normal module by Vite.

Reply:
@bluwy Understood! My reason against this: **we would be validating your Markdoc is correct at runtime.** This means, if you deploy to Astro SSR, you wouldn't see any validation errors until the first user makes a request 😬 Loading at buildtime ensures all Markdoc is valid _before_ deploying to the runtime. This mirrors the CI setup that Stripe uses for their deploys as well. I agree the esbuild loading isn't ideal, but this is the only solution to get the best of all worlds.

Reply:
I'll add a code comment to clarify!

Reply:
@bluwy I... still disagree with this, since validating when it's too late isn't really useful. The markdoc config is also at the base of your project outside of `src`, so it should be assumed that the same rules apply as the `astro.config` (no `import.meta.env` access, no aliases, no vite plugins respected).

I know this is a tough line to walk, but I feel pretty strongly about keeping buildtime validation for SSR. It would be worse to push to runtime so all-the-things work, then walk _back_ what you can do in a config if we hit problems later.

Reply:
Since markdoc is still in v0, let's go ahead with this then. I still have concerns with this technique but we can see how it goes 👍

Reply:
I pick the "validate as soon as possible" between the two evils. If there's a way to prevent errors from being reported to the end users, that should be the top priority. Developers should handle possible edge cases coming from esbuild/vite. Sure, it's a burden for the developers and us, but the end users shouldn't pay the cost.